File: solutions/go/pig-latin/pig_latin.go
package piglatin

import (
	"strings"
)

var vowels = map[byte]bool{'a': true, 'e': true, 'i': true, 'o': true, 'u': true}
var specials = map[string]bool{"xr": true, "yt": true}
var vowels_y = map[byte]bool{'a': true, 'e': true, 'i': true, 'o': true, 'u': true, 'y': true}

func Sentence(phrase string) string {
	result := strings.Builder{}

	for _, word := range strings.Split(phrase, " ") {
		if result.Len() != 0 {
			result.WriteByte(' ')
		}

		// If a word begins with a vowel, or starts with "xr" or "yt",
		// add an "ay" sound to the end of the word.
		if vowels[word[0]] || specials[word[:2]] {
			result.WriteString(word + "ay")
			continue
		}

		for pos := 1; pos < len(word); pos++ {
			letter := word[pos]
			// If a word starts with one or more consonants followed by "y",
			// first move the consonants preceding the "y"to the end of the word,
			// and then add an "ay" sound to the end of the word.
			if vowels_y[letter] {

				// If a word starts with zero or more consonants followed by "qu",
				// first move those consonants (if any) and the "qu" part to the end
				// of the word, and then add an "ay" sound to the end of the word.
				if letter == 'u' && word[pos-1] == 'q' {
					pos++
				}

				// If a word begins with one or more consonants, first move those consonants
				// to the end of the word and then add an "ay" sound to the end of the word.
				result.WriteString(word[pos:] + word[:pos] + "ay")
				break
			}
		}
	}
	return result.String()
}

File: solutions/lua/pig-latin/pig-latin.lua
return function(phrase)
    local function translate_word(word)
        local part1, part2 = nil, nil

        -- Rule 1: If a word begins with a vowel, or starts with "xr" or "yt"
        if word:find("^[aeiou]") or word:find("^xr") or word:find("^yt") then
            part1, part2 = word, ""
        end
        -- Rule3: If a word starts with zero or more consonants followed by "qu",
        -- move those consonants (if any) and the "qu" to the end of the word
        if not part1 then
            part1, part2 = word:match("^([^aeiou]*qu)(.*)") -- Rule 3
        end
        -- Rule 4: If a word starts with one or more consonants followed by "y",
        -- move the consonants preceding the "y"to the end of the word
        if not part1 then
            part1, part2 = word:match("^([^aeiou]+)(y.*)") -- Rule 4
        end
        -- Rule 2: If a word begins with one or more consonants,
        -- move those consonants to the end of the word
        if not part1 then
            part1, part2 = word:match("^([^aeiou]+)(.*)") -- Rule 2
        end
        if part1 and part2 then
            return string.format("%s%say", part2, part1)
        end
    end

    return phrase:gsub("%S+", translate_word)
end
